drivers/frequency/adf4371/adf4371.c

```diff
@@ -0,0 +1,820 @@
+	val = ADF4371_PD_POL(dev->pd_pol) |
+	      ADF4371_CP_CURRENT(dev->cp_settings.regval);
+
+	/* Set the phase detector polarity and the charge pump current */
+	ret = adf4371_update(dev, ADF4371_REG(0x1E),  mask, val);
+	if (ret < 0)
+		return ret;
+
+	/* Mute to Lock Detect */
+	if (dev->mute_till_lock_en) {
+		ret = adf4371_update(dev, ADF4371_REG(0x25),
+				     ADF4371_MUTE_LD_MSK,
+				     ADF4371_MUTE_LD(1));
+		if (ret < 0)
+			return ret;
+	}
+
+	ret = adf4371_update(dev, ADF4371_REG(0x20),
+			     ADF4371_MUXOUT_MSK |
+			     ADF4371_MUXOUT_EN_MSK |
+			     ADF4371_MUXOUT_LVL_MSK,
+			     ADF4371_MUXOUT(dev->muxout_default_mode) |
+			     ADF4371_MUXOUT_EN(dev->muxout_en) |
+			     ADF4371_MUXOUT_LVL(!dev->muxout_1v8_en));
+	if (ret < 0)
+		return ret;
+
+	/* Set address in ascending order, so the write_bulk() will work */
+	ret = adf4371_update(dev, ADF4371_REG(0x0),
+			     ADF4371_ADDR_ASC_MSK | ADF4371_ADDR_ASC_R_MSK,
+			     ADF4371_ADDR_ASC(1) | ADF4371_ADDR_ASC_R(1));
+	if (ret < 0)
+		return ret;
+
+	/*
+	 * Calculate and maximize PFD frequency
+	 * fPFD = REFIN × ((1 + D)/(R × (1 + T)))
+	 * Where D is the REFIN doubler bit, T is the reference divide by 2,
+	 * R is the reference division factor
+	 * TODO: it is assumed D and T equal 0.
+	 */
+	do {
+		dev->ref_div_factor++;
+		dev->fpfd = dev->clkin_freq / dev->ref_div_factor;
+	} while (dev->fpfd > ADF4371_MAX_FREQ_PFD);
+
+	/* Calculate Timeouts */
+	vco_band_div = DIV_ROUND_UP(dev->fpfd, 2400000U);
+
+	tmp = DIV_ROUND_CLOSEST(dev->fpfd, 1000000U);
+	do {
+		timeout++;
+		if (timeout > 1023) {
+			timeout = 2;
+			synth_timeout++;
+		}
+	} while (synth_timeout * 1024 + timeout <= 20 * tmp);
+
+	do {
+		vco_alc_timeout++;
+	} while (vco_alc_timeout * 1024 - timeout <= 50 * tmp);
+
+	dev->buf[0] = vco_band_div;
+	dev->buf[1] = timeout & 0xFF;
+	dev->buf[2] = ADF4371_TIMEOUT(timeout >> 8) | 0x04;
+	dev->buf[3] = synth_timeout;
+	dev->buf[4] = ADF4371_VCO_ALC_TOUT(vco_alc_timeout);
+
+	ret = adf4371_write_bulk(dev, ADF4371_REG(0x30), dev->buf, 5);
+	if (ret < 0)
+		return 0;
+
+	return adf4371_channel_config(dev);
+}
+
+/**
+ * Recalculate rate corresponding to a channel.
+ * @param dev - The device structure.
+ * @param chan - Channel number.
+ * @param rate - Channel rate.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t adf4371_clk_recalc_rate(struct adf4371_dev *dev, uint32_t chan,
+				uint64_t *rate)
+{
+	if (chan > dev->num_channels)
+		return FAILURE;
+
+	*rate = adf4371_pll_fract_n_get_rate(dev, chan);
+
+	return SUCCESS;
+}
+
+/**
+ * Calculate closest possible rate
+ * @param dev - The device structure
+ * @param rate - The desired rate.
+ * @param rounded_rate - The closest possible rate of desired rate.
+ * @return SUCCESS in case of success, negative error code otherwise.
+ */
+int32_t adf4371_clk_round_rate(struct adf4371_dev *dev, uint64_t rate,
+			       uint64_t *rounded_rate)
+{
+	*rounded_rate = rate;
```

Comment:
This API is a bit odd and it seems like it does not do much. Maybe this could be implemented by first saving `dev->integer`, `dev->fract1`, `dev->fract2` and `dev->mod2` in temporary variables and then doing: `adf4371_pll_fract_n_compute(rate, dev->fpfd, &dev->integer, &dev->fract1, &dev->fract2, &dev->mod2);
*rounded_rate = adf4371_pll_fract_n_get_rate(dev, channel);`
Followed by restoring the `dev` members to the original levels. But this would mean that the API should depend on the channel in some way.

Reply:
My question is: Is this API good as it is or does it need more work?